### PR TITLE
Allow user to exclude result from /actionexecutions, new endpoints for retrieving execution result, stdout and stderr

### PR DESCRIPTION
**Pull request by Kami**
## Changes

This pull request includes the following changes.

* Allow user to retrieve executions without the `result` attribute by passing `?exclude_result=1` query parameter to the `/actionexecutions/` endpoint.
* New endpoints:
  * `GET /actionexecutions/<id>/result` - retrieve execution result object
  * `GET /actionexecutions/<id>/stdout` - retrieve raw value of the execution result `stdout` attribute
  * `GET /actionexecutions/<id>/stderr` - retrieve raw value of the execution result `stderr` attribute

Other changes:

* Technical debt cleanup 
  * Removed awful pecan controller hackery and replaced it with nested controllers.
  * Tried to remove as much as the kwargs abuse and magical code as I could.

## Example responses

TBW

## Reasoning / Context

Currently, if an action generates a lot of output (`stdout`, `stderr`), retrieving all the actions executions will be very expensive. On top of that, JSON parsing of the response might fail due to the memory exhaustion if the object is too big (this happened in the ui when running in the browser).

This pull request solves this problem by allowing user to only retrieve action executions without the `result` attribute. In addition to that, it adds new endpoints for retrieving result, stdout and stderr attributes. This way, we can, for example inside the UI load the actions executions first and only request stdout and stderr when user hovers / clicks on the action execution (we can also pre-fetch some of that, etc., but that's a ui performance optimization).

This also goes along well with "result attribute needs to contain a result in a structured format" and the value of "stdout" and "stderr" is just an opaque (binary) data to us. I know this is currently not 100% true, but imo, we need to change and fix this in the near future (see https://github.com/StackStorm/st2/pull/1019#issuecomment-71622276).

This implementation is also very similar to the one I proposed for streaming (https://github.com/StackStorm/discussions/issues/27) and in the future, we can build streaming on top of those new endpoints.

Another thing related to the "stdout" and "stderr" being just an opaque binary data to us - currently we more or less use MongoDB for storing different meta data in it. This means storing potentially large "binary" / log structured data in it, in addition to the existing metadata is not the best fit (different access patterns, we don't need to query based on the value of this binary data, etc.)

In the future, we should (and probably will need to because of the performance reasons) look into storing this "binary" / log structured data in a different way / data store (e.g. even storing it in flat files might be better in some cases, but something like Cassandra is of course better than that).

### CODE REVIEW

File: st2common/st2common/models/db/__init__.py [hunk 1]
```diff
@@ -123,14 +83,24 @@ def get_all(self, *args, **kwargs):
     def count(self, *args, **kwargs):
         return self.model.objects(**kwargs).count()
 
-    @process_null_filter
-    @process_datetime_ranges
-    def query(self, *args, **kwargs):
-        offset = int(kwargs.pop('offset', 0))
-        limit = kwargs.pop('limit', None)
+    def query(self, offset=0, limit=None, order_by=None, exclude_fields=None,
```
Comment (Kami) [inline st2common/st2common/models/db/__init__.py]: This kwargs abuse and decorator magic was really awful. It's really hard to debug and follow what's going on. We should avoid kwargs abuse and such magic - remember, we are writing code for people and not for computers.

**Comment by enykeev** [comment]
Just want to add we also have a problem with `core.st2.generic.actiontrigger` exposing the whole execution including the results.

**Comment by Kami** [comment]
@enykeev You are referring to the trigger_instance attribute on the execution object, right?

**Comment by enykeev** [comment]
yeah, the initial execution gets exposed through `trigger_instance.payload`

**Comment by Kami** [comment]
@enykeev What do you think about handling it in a similar manner?

E.g. `?exclude_trigger_payload=1` and `/actionexecution/<id>/trigger_payload`?

**Comment by manasdk** [comment]
I dont agree with these endpoints -

GET /actionexecutions/<id>/stdout - retrieve raw value of the execution result stdout attribute
GET /actionexecutions/<id>/stderr - retrieve raw value of the execution result stderr attribute

stdout and stderr are properties specific to runners. Exposing them here feels broken. If the results endpoint was able to handle retrieval of a specific property then it makes sense.

Basically, we are looking to support projections of an object. Can it be done generically rather than these specific endpoints?

**Comment by Kami** [comment]
@manasdk Yeah, I was thinking about that (being specific to a runner). In the streaming endpoint, I proposed an endpoint called "output" which would combine stdout and stderr, but we still want potentially retrieve it separately in this case.

Something like this could work:

```
GET /actionexecutions/<id>/result/<attribute name>
```

To be on the safe side, we should still limit the attributes to something like stdout, stderr, trigger.payload.

**Comment by enykeev** [comment]
what are we going to do with runners like `run-remote` that don't have stdout in the root of the results instead having more complex structure?

**Comment by Kami** [comment]
@enykeev That's a good point - one option is to have a per-runner specific processing, but I'm not a big fan of what.

Other option is to have per-runner specific endpoints, e.g. for remote runner, something like `GET /actionexecutions/<id>/result/<hostname>/<attribute name>/`. Not a huge fan of that either.

**Comment by Kami** [comment]
Alright, so we had a pretty heated (:smiling_imp:)  debate about this on Slack.

We decided to go with the following approach / changes:

* `GET /actionexecutions/?exclude_attributes=result,trigger.instance` - allow user to exclude particular attributes from the response when listing all the executions
* `GET /actionexecutions/<id>?exclude_attributes=result,trigger_instance` - allow user to exclude particular attributes from the response when retrieving details for a single execution
* `GET /actionexecutions/<id>/attribute/<{result,trigger_instance}>` - allow user to retrieve a value of a particular attribute for a particular execution

Just first change should hopefully get us far, but second and third change are an additional "bonus".

I'm working on those changes as we speak. @enykeev I'll ping you when it's ready for you to hook it up in the UI.

**Comment by Kami** [comment]
Also, to clarify it, when I get a green light, I will merge those changes into `master` and `0.8.1` branch.

**Comment by Kami** [comment]
@enykeev Alright, changes are ready for you. Please let me know when you have tested the changes and hooked them up in the UI and I will get them merged into 0.8.1 and master.

**Comment by enykeev** [comment]
@Kami LGTM. Changes in st2web are made, although it would be great if we could test them in an actual problematic environment before pushing 0.8.1. WDYT, is there any way for us to deploy them on master002 or its exact copy to make sure it actually solves the problem?

Feel free to merge StackStorm/st2web#108 as soon as this PR lands.

**Comment by Kami** [comment]
@enykeev Thanks - I'll try to coordinate this with @DoriftoShoes.

In any case, merging into master is fine in both case (we should just wait with merging it into 0.8 branch :))

**Comment by Kami** [comment]
I also checked out the st2web branch and tested it against this branch and it looks good.

In the future we could also look in exclude result and trigger instance being a default behavior - maybe we can do this change in 0.9 and in 0.8 we keep it backward compatible and as it is now.

**Comment by Kami** [comment]
Merged changes into `0.8.1` and `master` branch.
